Fix help(latex) to show the docstring

asmeurer opened the pull request:
Previously, `functools.update_wrapper()` was used on the wrapper class instance.
But this added `__doc__` to the class instance, which does not work with `help()`
(`help()` only looks at `__doc__` if it is defined on the class). So we instead
redefine the wrapper class for each decorated function, and use `update_wrapper`
on the class rather than the instance).

Fixes #20231.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

(note, @oscarbenjamin if we want to backport this to the release branch, we don't need a release notes entry as I believe this was a regression since the last release)

eric-wieser commented on sympy/printing/printer.py:
```diff
@@ -347,39 +347,39 @@ def _as_ordered_terms(self, expr, order=None):
-
 def print_function(print_cls):
     """ A decorator to replace kwargs with the printer settings in __signature__ """
-    def decorator(f):
-        return _PrintFunction(f, print_cls)
-    return decorator
```
How about localizing the hack to right here?
```python
def decorator(f):
    # We have to create a subclass so that `help` actually shows the docstring. IPython and Sphinx do not
    # need this, only a raw python terminal.
    cls = type(f'{f.__qualname__}_doc_hack', (_PrintFunction,), dict(__doc__=f.__doc__))
    return cls(f, print_cls)
```

eric-wieser replied:
As remarked in https://github.com/sympy/sympy/issues/20231#issuecomment-715897919, this is only necessary in Python 3.6, 3.7, and 3.8.

asmeurer replied:
That is probably fine. I haven't looked at if any of the other attributes updated in update_wrapper are broken from doing it on a class instance. It's good to see that Python 3.9 actually does the right thing here now. 

Actually, perhaps a bigger concern is that in both cases, we are doing this at import time, so it's not great in terms of keeping the import time down. Is there really no way to monkeypatch the `__signature__` of a function?

asmeurer replied:
I guess it's so that changing the global defaults changes the signature. I'm not really sure if this is worth all this metaprogramming.

eric-wieser replied:
> Is there really no way to monkeypatch the __signature__ of a function?

Not if we want it to update to reflect the global defaults set by init_printing

> Actually, perhaps a bigger concern is that in both cases, we are doing this at import time, so it's not great in terms of keeping the import time down.

I don't think class creation is expensive, so neither patch should make a difference over master. The expensive bit is the sanity check in `__init__` that calls inspect.signature, which could be skipped.

eric-wieser replied:
> I'm not really sure if this is worth all this metaprogramming.

I'd argue that the reuse across lots of printers makes this worthwhile.

asmeurer replied:
OK. It's really unfortunate that the `__signature__` API uses an attribute instead of a method call. I've run into this problem in other projects, where you have to do some metaprogramming to make `__signature__` give a dynamic thing (the good news is to do so on a class only requires using a descriptor, but that won't work for functions like `latex`).

asmeurer commented on sympy/printing/printer.py:
```diff
@@ -347,39 +347,39 @@ def _as_ordered_terms(self, expr, order=None):
+            self.__print_cls = print_cls
+            update_wrapper(type(self), f, updated=())
+
+        def __repr__(self) -> str:
```
By the way, I think we shouldn't try to change the repr for this class. I don't think it's a big deal for users to see that the functions are actually wrapper functions. On the other hand, making the repr look like a function can make things very confusing. I was quite confused when debugging this until I noticed this method defined here.

eric-wieser replied:
No objection from me. I did it this way because it was more useful than the default repr, and easier that deciding on a better one.

eric-wieser commented:
In case you missed it, I introduced this code in #20067

asmeurer commented:
I updated this with @eric-wieser's suggestion, and removed `__repr__` from `_PrintFunction`.

eric-wieser commented on sympy/printing/printer.py:
```diff
@@ -360,9 +361,6 @@ def __init__(self, f, print_cls: Type[Printer]):
         update_wrapper(self, f)
 
-    def __repr__(self) -> str:
-        return repr(self.__wrapped__)  # type:ignore
```
I think it would be good if `repr(latex)` contained either the word `latex`, `LatexPrinter`, or both - from that perspective, removing the repr entirely makes things worse.

asmeurer replied:
It does in Python 3.8:

```py
>>> latex
<sympy.printing.printer.latex_function object at 0x7fe1f9ac0110>
```

We can dynamically create the class in Python 3.9 as well if this seems important.

eric-wieser replied:
For the sake of hiding the <3.9 hack, I'd be tempted to go with
```suggestion
    def __repr__(self):
        return f"_PrintFunction({self.__print_cls!r}, {self.__wrapped__!r})"

```
and the change I suggest below...

eric-wieser commented on sympy/printing/printer.py:
```diff
@@ -381,5 +379,11 @@ def __signature__(self) -> inspect.Signature:
+        if sys.version_info < (3, 9):
+            # We have to create a subclass so that `help` actually shows the docstring in older python versions.
+            # IPython and Sphinx do not need this, only a raw python console.
+            cls = type(f'{f.__qualname__}_PrintFunction', (_PrintFunction,), dict(__doc__=f.__doc__))
```
```suggestion
            cls = type(f'_PrintFunction', (_PrintFunction,), dict(__doc__=f.__doc__))
```
(hiding the class name since the subclass is a hack)

asmeurer replied:
I guess I'm not keen on hiding the fact that we have this hack. It makes the debugging harder. 

By the way, I hope this dynamically created subclass doesn't cause issues with things like pickling. If it does, then I think the whole thing really isn't worth it, and we should just use the non-updating version of assigning `__signature__` to the function.

eric-wieser replied:
So, there are two different "hacks" here:

* The existence of `_PrintFunction`, which is to ensure that `latex` actually respects the global settings and that `inspect.signature` shows them. Previously we did only the former, then someone didn't like the documentation was broken and fixed the latter at the expense of breaking the former.
* The existence of the subclass in Python < 3.8, which is just to fix the builtin `help`

IMO, the former hack is interesting, and the latter is not - so it's reasonable to try and hide the parts of the latter hack which are just noise.

> I hope this dynamically created subclass doesn't cause issues with things like pickling

Looks like I already broke that. It's easy to fix with:
```
def __reduce__(self):
    return self.__wrapped__.__name__
```

eric-wieser replied:
xref #20468, where I fix pickling

eric-wieser reviewed (approved):
What you have looks good, but some suggested `repr` changes.

czgdp1807 commented:
@asmeurer @eric-wieser Is this PR ready to be merged?

tqa236 commented:
Hello all, is this PR ready to be merged?

czgdp1807 commented:
@asmeurer There are some conflicts. Please resolve them.

oscarbenjamin committed:
Fix help(latex) to show the docstring

Previously, functools.update_wrapper() was used on the wrapper class instance.
But this added __doc__ to the class instance, which does not work with help()
(help() only looks at __doc__ if it is defined on the class). So we instead
redefine the wrapper class for each decorated function, and use update_wrapper
on the class rather than the instance).

Use a dynamically created type for _PrintFunction in Python <3.8

This is simpler than moving the class definition to the body of the function.
__doc__ does not need to be on the class itself in Python 3.9 for it to work
with help(), but it does for 3.8 and lower. This alternative change was
suggested by @eric-wieser.

This also removes __repr__ from the class, as it only serves to make things
confusing when debugging this class, and there's no reason to completely hide
the fact that it is a class instead of the function.

Co-authored-by: Aaron Meurer <asmeurer@gmail.com>
Co-authored-by: Oscar Gustafsson <oscar.gustafsson@gmail.com>

oscarbenjamin commented:
I also opened #22038 to apply this in the 1.9 branch and I've moved the release note over there.
